### PR TITLE
Feature/openssl digest

### PR DESCRIPTION
**Pull request by benoist**

### CODE REVIEW

File: src/openssl/digest/digest.cr [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    getter name
+
+    def initialize(@name, @ctx : LibCrypto::EVP_MD_CTX)
+      raise DigestError.new("Invalid EVP_MD_CTX") unless @ctx
+    end
+
+    def self.hexdigest(data)
+      raise DigestError.new("Cannot digest")
+    end
```
Comment (jhass) [inline src/openssl/digest/digest.cr]: Could just define this as abstract?

File: src/openssl/digest.cr [hunk 2]
```diff
@@ -0,0 +1,21 @@
+require "./lib_crypto"
+require "./digest/*"
+class OpenSSL::Digest
+  def self.digest(algorithm)
+    case algorithm
+    when :dss       then DSS.new
+    when :dss1      then DSS1.new
+    when :md4       then MD4.new
+    when :md5       then MD5.new
+    when :mdc2      then MDC2.new
```
Comment (jhass) [inline src/openssl/digest.cr]: > MDC-2 was covered by U.S. Patent 4,908,861, issued on March 13, 1990 but filed by IBM on August 28, 1987.
For this reason, support for MDC-2 has been disabled in OpenSSL on most Linux distributions and is not implemented by many other cryptographic libraries
https://en.wikipedia.org/wiki/MDC-2#Patent_issues

And that's failing the linux build apparently.

**Comment by technorama** [comment]
@jhass @benoist It would be better to drop the extra classes and always use the string initializer.
* OpenSSL adds and removes Ciphers/Digests on occasion between releases.
* OpenSSL forks may add or remove support for certain Ciphers/Digests.
* OpenSSL may be compiled without support for certain ciphers.
* Ciphers and Digests can be added at runtime (custom ciphers).
* Cipher/Digest names can be stored in files or sent over a network.

What happens when the cipher isn't supported on a particular build of OpenSSL?  The class exists but can't be instantiated.   That seems like lying to the user.  With c you can use `#ifdef` to determine if the cipher is supported.  Not so here.

The string form should be the preferred and only way to instantiate the class even if the class form looks slightly nicer.

**Comment by benoist** [comment]
@technorama but by using the string initializer it will always cause runtime errors instead of compile time errors when trying to use an invalid Digest.
Having said that, I do think we should be able to user of the library to add and remove digest methods. A user would already be able to do so via the def_digest_classes(names) macro.

**Comment by technorama** [comment]
That's not completely true.  A missing cipher that's either unsupported in the current version or that was compiled without support is still a runtime error even if the class exists.  Having the class is a lie.  It may not work.  At least with the string version you can expect it to raise because of the explicit string parameter.

Removing string instantiation also removes the possibility to store the cipher names elsewhere (files, network) which is something I often do to provide an upgrade path for weak or broken ciphers/digests. 

In ruby there was a lot of lag between an OpenSSL release and ruby updating it's OpenSSL library to support new features or API changes (it broke several times).  LibreSSL and BoringSSL are deprecating and/or adding digests and ciphers.  Who is going to keep that list up to date?  Will crystal releases always happen every few months or will it fit in to a yearly schedule (or longer) like ruby?  I expect @asterite to tell me crystal is in beta and under fast development but that's not my point.  What happens when crystal is stable and can't make new releases just to accommodate OpenSSL/LibreSSL/BoringSSL breakage?  Who is planning for the future with the realities of users complaining about crystal failures because their platform uses a newer SSL version?  The ruby community periodically goes through this problem and is planning in gemifying OpenSSL to remove that dependency.  If we were smart we'd recognize the problems with putting OpenSSL in the base and that others (ruby) are attempting to rectify it by separating OpenSSL from the main ruby distribution.  If there was a way to override the builtin OpenSSL with shards we could avoid the problems that ruby had.  A default set of shards provided with crystal that can be overridden would (probably) solve the issue.

**Comment by technorama** [comment]
Here's a list of ciphers taken from the documentation.  Do you want this many classes?  I thought crystal had problems with lots of classes.  Perhaps this isn't enough to cause an issue.

```
 bf-cbc             Blowfish in CBC mode
 bf                 Alias for bf-cbc
 bf-cfb             Blowfish in CFB mode
 bf-ecb             Blowfish in ECB mode
 bf-ofb             Blowfish in OFB mode

 cast-cbc           CAST in CBC mode
 cast               Alias for cast-cbc
 cast5-cbc          CAST5 in CBC mode
 cast5-cfb          CAST5 in CFB mode
 cast5-ecb          CAST5 in ECB mode
 cast5-ofb          CAST5 in OFB mode

 des-cbc            DES in CBC mode
 des                Alias for des-cbc
 des-cfb            DES in CBC mode
 des-ofb            DES in OFB mode
 des-ecb            DES in ECB mode

 des-ede-cbc        Two key triple DES EDE in CBC mode
 des-ede            Two key triple DES EDE in ECB mode
 des-ede-cfb        Two key triple DES EDE in CFB mode
 des-ede-ofb        Two key triple DES EDE in OFB mode

 des-ede3-cbc       Three key triple DES EDE in CBC mode
 des-ede3           Three key triple DES EDE in ECB mode
 des3               Alias for des-ede3-cbc
 des-ede3-cfb       Three key triple DES EDE CFB mode
 des-ede3-ofb       Three key triple DES EDE in OFB mode

 desx               DESX algorithm.

 gost89             GOST 28147-89 in CFB mode (provided by ccgost engine)
 gost89-cnt        `GOST 28147-89 in CNT mode (provided by ccgost engine) 

 idea-cbc           IDEA algorithm in CBC mode
 idea               same as idea-cbc
 idea-cfb           IDEA in CFB mode
 idea-ecb           IDEA in ECB mode
 idea-ofb           IDEA in OFB mode

 rc2-cbc            128 bit RC2 in CBC mode
 rc2                Alias for rc2-cbc
 rc2-cfb            128 bit RC2 in CFB mode
 rc2-ecb            128 bit RC2 in ECB mode
 rc2-ofb            128 bit RC2 in OFB mode
 rc2-64-cbc         64 bit RC2 in CBC mode
 rc2-40-cbc         40 bit RC2 in CBC mode

 rc4                128 bit RC4
 rc4-64             64 bit RC4
 rc4-40             40 bit RC4

 rc5-cbc            RC5 cipher in CBC mode
 rc5                Alias for rc5-cbc
 rc5-cfb            RC5 cipher in CFB mode
 rc5-ecb            RC5 cipher in ECB mode
 rc5-ofb            RC5 cipher in OFB mode

 aes-[128|192|256]-cbc  128/192/256 bit AES in CBC mode
 aes[128|192|256]       Alias for aes-[128|192|256]-cbc
 aes-[128|192|256]-cfb  128/192/256 bit AES in 128 bit CFB mode
 aes-[128|192|256]-cfb1 128/192/256 bit AES in 1 bit CFB mode
 aes-[128|192|256]-cfb8 128/192/256 bit AES in 8 bit CFB mode
 aes-[128|192|256]-ecb  128/192/256 bit AES in ECB mode
 aes-[128|192|256]-ofb  128/192/256 bit AES in OFB mode
```

**Comment by benoist** [comment]
I've been coding in Ruby for a long time and I'm really comfortable in expressing my opinion on design decisions in that language. I've only recently started working with crystal so I'm not really comfortable in deciding if more classes is good or not. 
The pull requests for Digest and Cipher contain code from the openssl.cr shard so the decisions in design are not really mine. I've started presenting the small parts from the original shard to start a discussion about the implementation. 
I'm happy to make changes and make them acceptable for merging.

> That's not completely true. A missing cipher that's either unsupported in the current version or that was compiled without support is still a runtime error even if the class exists.

Yeah I've learned that the hard way as well since the build is failing because of this :-)

Also looking at the list of Ciphers I don't think its a good idea the have classes for each of them.

**Comment by ysbaddaden** [comment]
The String only is less code and should be a bit stronger against *SSL discrepancies, and this should be the most stable over time. Maybe we should use this to being with.

Having many classes isn't a problem as long as they are required on demand, and not always. Thought it doesn't really make sense to have classes for just a few selected ciphers and not all of them.

**Comment by bararchy** [comment]
@technorama Most of those ciphers are deprecated or not used, we should not add all of them, only relevant ones.

To-Remove:

rc2, gost, idea, cast.

all the other ones should be used.

**Pull request closed without merging**